### DISCUSSION TITLE
Rule for verifying parent relation ownership

### DISCUSSION THREAD

**Discussion by t6d** [pocketbase/pocketbase#427] [Q&A] [answered]
I'm just playing around with a todo list example to get a better understanding of the API. I have the following schema with `tasks.list` and `lists.user` being defined as relations:

```sql
CREATE TABLE `lists` (
  `created` TEXT DEFAULT "" NOT NULL, 
  `id` TEXT PRIMARY KEY, 
  `name` TEXT DEFAULT '', 
  `updated` TEXT DEFAULT "" NOT NULL, 
  "user" TEXT DEFAULT ''
);
CREATE INDEX `lists_created_idx` ON `lists` (`created`);
```

and

```sql
CREATE TABLE `tasks` (
  `completed` Boolean DEFAULT FALSE, 
  `created` TEXT DEFAULT "" NOT NULL, 
  "list" TEXT DEFAULT '', 
  `id` TEXT PRIMARY KEY, 
  `summary` TEXT DEFAULT '', 
  `updated` TEXT DEFAULT "" NOT NULL
);
CREATE INDEX `tasks_created_idx` ON `tasks` (`created`);
```

How do I create a rule that

a.) ensures that a request to create a task always contains a list and
b.) that the user is the owner of the list the task references?

**Comment by ganigeorgiev** [comment, marked as answer]
You can't create the underlying collections table only by raw `CREATE TABLE` statements because PocketBase needs additional meta data to operate, aka. the collection, to perform access checks, validations and apply constratints.

If you want to create collections programmatically (that will automatically create its underlying table), you can make use of the [Collection API](https://pocketbase.io/docs/api-collections/) or use [PocketBase as framework](https://pocketbase.io/docs/db-queries/) and directly interact with the collections DAO methods (there will be more documentation on that in the near future).

To ensure that `list` field in your example always contains a `list` reference, you need to create a `relation` type field pointing to the `lists` collection and mark it "Required".

To allow for example listing `tasks` that are linked to `lists` owned by the current logged in user, you can set the following collection api rule in the `tasks` collection:
```
list.user = @request.user.id
```

More information on the API rules you could find in https://pocketbase.io/docs/manage-collections/#collection-api-rules.

**Reply by ganigeorgiev** [reply]
_Please also note that the above will slightly change in the near future because of the ongoing users management refactoring in #376._

**Reply by t6d** [reply]
> You can't create the underlying collections table only by raw CREATE TABLE statements because PocketBase needs additional meta data to operate, aka. the collection, to perform access checks, validations and apply constratints.

I did that. I only supplied the SQL for illustration purposes. :) 

> To ensure that list field in your example always contains a list reference, you need to create a relation type field pointing to the lists collection and mark it "Required".

I didn't realize it is as simple as checking that the list belong to the user through the relation. I feel like I complicated that in my head much more.

Thanks for the quick help!

**Reply by t6d** [reply]
I think I now know why I had this wrong yesterday. It wasn't clear to me that I can operate on the record that is about to be inserted. I was trying to operate on `@request.data` thinking of it more as a pre-condition that runs before insert. Hence, I was wondering how to then relate that request data to the records in the database. It's obviously much simpler, as you outlined above, to operate directly on the record being inserted. Maybe that's worth calling out in the documentation.

While I'm here, I just want to say that this is an amazing product and a great use of Go's ability to create zero-dependency binaries. On top of that, the Admin UI looks really snazzy! Great work! 👨‍🎨 ✨

**Reply by ganigeorgiev** [reply]
> I think I now know why I had this wrong yesterday. It wasn't clear to me that I can operate on the record that is about to be inserted.

Yep, this will be outlined in the future documentation changes planned in #376 (and probably in the UI too).

In general:

- `@request.data.*` contains the request submitted body data.

- the access rules for `list`, `view`, `update`, `delete` operates over the already existing record, or in other words `name != ""` will check whether the existing stored `name` value is not empty.

- the `create` rule is a little different from the above ones because we don't have an existing record to get the value from and therefore the rule is applied after a transactional dry-submit (without files upload/delete). If it fails - we just rollback the statement. Or in other words, if we have a rule like `name != ""` for create it will be an alias for `@request.data.name != ""` (we populate the data in the record because we want to allow querying nested referenced relation fields, eg. `rel.field1.field2 > 100`).
